audiobookshelf: Implement more efficient multi-file seeking

tsipinakis opened the pull request:
The previous seeking implementation relied on downloading all audiobook files and concatenating them together (and then doing it again every time the user seeked). With a large book, this very quickly became slow and a resource hog.

A more efficient way is to use the audio track duration information to determine which track to start the playback at.

OzGav commented:
@fmunkes FYI

tsipinakis commented:
(There's no mark as draft button on GH?)

Anyway, reviews are welcome. I'm currently using this at home to chase out any remaining bugs. I'll let you know when I consider this bug-free.

If you see anything obvious in the code feel free to comment :)

fmunkes commented:
> (There's no mark as draft button on GH?)
> 
> Anyway, reviews are welcome. I'm currently using this at home to chase out any remaining bugs. I'll let you know when I consider this bug-free.
> 
> If you see anything obvious in the code feel free to comment :)

Take your time, no need to hurry.

tsipinakis commented:
Ok I think this is ready to review now - managed to find some time to field test it today :)
Listened to quite a few chapters & seeked around, no issues found.

tsipinakis commented on music_assistant/providers/audiobookshelf/__init__.py:
```diff
@@ -526,6 +525,66 @@ async def _get_stream_details_audiobook(
+        self, tracks: list[AbsAudioTrack], seek_position: int
+    ) -> tuple[int, int]:
+        for i, track in enumerate(tracks):
+            self.logger.debug(f"Seeking to position {seek_position}")
```
This is useful for comparing with the requested position. Up to you what level of debug logging you want to keep.

fmunkes commented on music_assistant/helpers/audio.py:
```diff
@@ -77,6 +77,8 @@
 CACHE_FILES_IN_USE: set[str] = set()
 
+__all__ = ["get_multi_file_stream"]
+
```
Why do we need the `__all__` here ? The shouldn't be a `from ... import *` anywhere.

tsipinakis replied:
Indeed I'm not sure why that's there, removed.

fmunkes commented on music_assistant/helpers/audio.py:
```diff
@@ -1287,6 +1289,7 @@ async def get_file_stream(
 async def get_multi_file_stream(
     mass: MusicAssistant,  # noqa: ARG001
     streamdetails: StreamDetails,
+    seek_position: int = 0,
```
Please add a unit comment here.

tsipinakis replied:
Done.

fmunkes commented on music_assistant/providers/audiobookshelf/__init__.py:
```diff
@@ -526,6 +527,72 @@ async def _get_stream_details_audiobook(
         )
 
+    def _get_track_from_position(
+        self, tracks: list[AbsAudioTrack], seek_position: int
```
Another unit for me please - or rename it to something like "seek_position_s" to include the unit.

tsipinakis replied:
Done.

fmunkes commented on music_assistant/providers/audiobookshelf/__init__.py:
```diff
@@ -526,6 +527,72 @@ async def _get_stream_details_audiobook(
+        self, tracks: list[AbsAudioTrack], seek_position: int
+    ) -> tuple[int, int]:
+        for i, track in enumerate(tracks):
+            self.logger.debug(f"Seeking to position {seek_position}")
```
Maybe we can shorten the debug log to something like "skipped x tracks to seek to position xy." Otherwise possible error reports of users might be hard to read.

tsipinakis replied:
Done

fmunkes commented on music_assistant/providers/audiobookshelf/__init__.py:
```diff
@@ -526,6 +527,72 @@ async def _get_stream_details_audiobook(
+                f"Found seeked track {track.title}, offset {offset}, position {position}"
+            )
+            return i, position
+        return 0, 0
```
I think the index is hard to read - what about return the remaining list of tracks directly, and there position?

tsipinakis replied:
Done, now returns a track list and position

fmunkes commented on music_assistant/providers/audiobookshelf/__init__.py:
```diff
@@ -526,6 +527,72 @@ async def _get_stream_details_audiobook(
+        self, streamdetails: StreamDetails, seek_position: int = 0
+    ) -> AsyncGenerator[bytes, None]:
+        """Retrieve the audio track at the requested position."""
+        i, position = self._get_track_from_position(streamdetails.data, seek_position)
```
See above. If you'd like to keep the index, then please name it track_index or similar.

tsipinakis replied:
Fixed

fmunkes reviewed (changes requested):
That works nicely! Just added a couple of comments, which should ease future maintainability I think.

marcelveldt commented:
> The previous seeking implementation relied on downloading all audiobook files and concatenating them together (and then doing it again every time the user seeked).

That is not exactly true as it would download and concatenate the files into a cache for fast local seeking. 
So maybe that in your case there is not enough diskspace to do this local caching.

In any case, a more efficient seeking within the provider itself is always preferred over having to download and concatenate all parts

tsipinakis committed:
audiobookshelf: Implement more efficient multi-file seeking

The previous seeking implementation relied on downloading all audiobook
files and concatenating them together (and then doing it again every
time the user seeked). With a large book, this very quickly became slow
and a resource hog.

A more efficient way is to use the audio track duration information to
determine which track to start the playback at.

OzGav commented:
@tsipinakis you can mark the PR comments as resolved where you indicate you have fixed the comment. You can decide whether the AI suggestion is valid and resolve that as well. Once all suggestions are resolved you can request another review from Fabian.

edit: also update the branch to make sure there are no introduced problems there.

fmunkes commented:
Excellent! Thanks @tsipinakis !
